code/krepel/memory/allocation.d

```diff
@@ -0,0 +1,725 @@
+  {
+    assert(Byte == 0xFU);
+  }
+
+  assert(SystemHeap.Deallocate(Block1));
+  foreach(ref Byte; Block2)
+  {
+    assert(Byte == 0xFU);
+  }
+
+  assert(SystemHeap.Deallocate(Block2));
+  //debug {} else assert(!SystemHeap.Deallocate(Block2));
+
+  auto Block3 = SystemHeap.Allocate(9, 16);
+  assert(Block3);
+  assert(Block3.ptr == AlignedPointer(Block3.ptr, 16));
+}
+
+// Heap Memory Allocation
+unittest
+{
+  ubyte[256] Buffer = void;
+  for(int Index; Index < Buffer.length; Index++)
+  {
+    Buffer[Index] = cast(ubyte)Index;
+  }
+
+  auto Heap = HeapMemory(Buffer[]);
+
+  auto Block1 = Heap.Allocate(16);
+  assert(Block1.length == 16);
+  auto Block2 = Heap.Allocate(16);
+  assert(Block2.ptr != Block1.ptr);
+  assert(Block2.length == 16);
+
+  // TODO(Manu): Test reliably for the out-of-memory case somehow.
+  //assert(Heap.Allocate(1) is null);
+
+  assert(Heap.Deallocate(Block1));
+
+  auto Block3 = Heap.Allocate(16);
+  assert(Block3.ptr == Block1.ptr);
+  assert(Block3.length == 16);
+}
+
+// Dynamic stack allocation
+unittest
+{
+  ubyte[1024] Buffer = void;
+
+  StackMemory Stack;
+  // Only assign 128 bytes as memory.
+  Stack.Initialize(Buffer[0 .. 128]);
+  assert(Stack.AllocationMark == 0);
+
+  auto Block1 = Stack.Allocate(32);
+  assert(Block1.length == 32);
+  assert(Stack.AllocationMark == 32);
+  for(size_t Index = 0; Index < Block1.length; Index++)
+  {
+    assert(Block1[Index] == Buffer[Index]);
+  }
+
+  auto Block2 = Stack.Allocate(64);
+  assert(Block2.length == 64);
+  assert(Block2.ptr == Block1.ptr + Block1.length);
+  assert(Stack.AllocationMark == 32 + 64);
+  for(size_t Index = 0; Index < Block2.length; Index++)
+  {
+    assert(Block2[Index] == Buffer[Block1.length + Index]);
+  }
+
+  // There's still room for a bit more, but 512 will not fit.
+  auto Block3 = Stack.Allocate(512);
+  assert(Block3 is null);
+  assert(Stack.AllocationMark == 32 + 64, "Allocation mark should be unchanged.");
+}
+
+// Static stack allocation
+unittest
+{
+  StaticStackMemory!128 Stack;
+
+  auto Block1 = Stack.Allocate(32);
+  assert(Block1);
+  assert(Stack.AllocationMark == 32);
+
+  auto Block2 = Stack.Allocate(64);
+  assert(Block2);
+  assert(Stack.AllocationMark == 32 + 64);
+
+  auto Block3 = Stack.Allocate(128);
+  assert(Block3 is null);
+  assert(Stack.AllocationMark == 32 + 64);
+}
+
+// HybridMemory tests
+unittest
+{
+  ubyte[256] Buffer;
+  auto Heap1  = HeapMemory(Buffer[  0 .. 128]);
+  auto Heap2  = HeapMemory(Buffer[128 .. 256]);
+  auto Hybrid = HybridMemory!(HeapMemory*, HeapMemory*)(&Heap1, &Heap2);
+
+  auto Block1 = Hybrid.Allocate(16);
+  assert(Block1);
+  assert(Block1.length == 16);
+  assert(Heap1.Contains(Block1));
+  auto Block2 = Hybrid.Allocate(16);
+  assert(Block2.length == 16);
+  assert(Heap1.Contains(Block1));
+  auto Block3 = Hybrid.Allocate(16);
+  assert(Block3.length == 16);
+  assert(Heap1.Contains(Block1));
+  auto Block4 = Hybrid.Allocate(16);
+  assert(Block4);
```

Comment:
No test for the actual hybrid stuff? You exactly stop, when trying to do allocate from the second block.